rcon/commands.py

```diff
@@ -73,78 +89,119 @@ class ServerCtl:
     set password not implemented on purpose
     """
 
-    def __init__(self, config, auto_retry=1):
+    def __init__(self, config, auto_retry=1, pool_size=20):
         # .env fed config from rcon.SERVER_INFO
         self.config = config
-        self.conn = None
-        # self._connect()
         self.auto_retry = auto_retry
+        self.mu = threading.RLock()
+        self.idles: list[HLLConnection] = []
+        self.numOpen = 0
+        self.maxOpen = 20
+        self.maxIdle = pool_size
+
+    @contextmanager
+    def with_connection(self) -> HLLConnection:
+        logger.debug("Waiting to acquire lock")
+        with self.mu:
+            if len(self.idles) != 0:
+                logger.debug("acquiring connection from idle pool")
+                conn = self.idles.pop()
+
+            elif self.numOpen >= self.maxOpen:
+                logger.debug("Max connections already open, waiting for connection returned to pool")
+                c = 0
+                while len(self.idles) == 0:
+                    if c >= 30:
+                        logger.error("waiting for connection returned to pool timed out after " + str(c) + " seconds")
+                        raise TimeoutError()
+                    c += 1
+                    time.sleep(1)
+                conn = self.idles.pop()
+
+            else:
+                logger.debug("Opening a new connection")
+                conn = HLLConnection()
+                self._connect(conn)
+                self.numOpen += 1
 
-    def _connect(self):
-        self.conn = HLLConnection()
         try:
-            self.conn.connect(
-                self.config["host"], int(self.config["port"]), self.config["password"]
-            )
+            yield conn
+        finally:
+            logger.debug("return connection")
+            if len(self.idles) >= self.maxIdle:
+                logger.debug("Enough connections in pool, closing")
+                self.numOpen -= 1
+                conn.close()
+            else:
+                logger.debug("Returning connection to pool")
+                self.idles.append(conn)
+
+    def _connect(self, conn: HLLConnection):
+        try:
+            conn.connect(self.config["host"], int(self.config["port"]), self.config["password"])
         except (TypeError, ValueError) as e:
-            logger.critical("Invalid connection information")
+            logger.critical("Invalid connection information", e)
```

Comment:
you need to add the %s in the format string also. but if you use `logger.exception` the stacktrace should contain the exception name an its content anyways. Dunno why I used cirtical in there, not sure how it behaves actually

Reply:
lol, you might be right, didn't really look into that :D I'll might simply change that to logger.exception 😅